src/ng/directive/ngAs.js

```diff
@@ -0,0 +1,148 @@
+ *     angular.module('ngAsVoteExample', [])
+ *      .component('voteTaker', {
+ *         controller: function() {
+ *           this.vote = null;
+ *           this.agree = function() { this.vote = 'agree'; };
+ *           this.disagree = function() { this.vote = 'disagree'; };
+ *           this.clear = function() { this.vote = null; };
+ *         },
+ *         template:
+ *           '<button ng-disabled="$ctrl.vote" ng-click="$ctrl.agree()">Agree</button>' +
+ *           '<button ng-disabled="$ctrl.vote" ng-click="$ctrl.disagree()">Disagree</button>'
+ *       })
+ *       .component('competition', {
+ *          controller: function() {
+ *            this.redVoteTaker = null;
+ *            this.blueVoteTaker = null;
+ *            this.match = function() {
+ *              return this.redVoteTaker.vote === this.blueVoteTaker.vote && this.redVoteTaker.vote;
+ *            };
+ *            this.next = function() {
+ *              this.sentence++;
+ *              this.redVoteTaker.clear();
+ *              this.blueVoteTaker.clear();
+ *            };
+ *          },
+ *          template:
+ *            '<p>Red team: <vote-taker ng-as="$ctrl.redVoteTaker"></vote-taker></p>' +
+ *            '<p>Blue team: <vote-taker ng-as="$ctrl.blueVoteTaker"></vote-taker></p>' +
+ *            '<p ng-show="$ctrl.match()">There is a match!</p>' +
+ *            '<button ng-click="$ctrl.next()">Next</button>'
+ *       });
+ *   </file>
+ *   <file name="protractor.js" type="protractor">
+ *      var agrees = element.all(by.buttonText('Agree'));
+ *      var matchMessage = element(by.css('[ng-show]'));
+ *      var next = element(by.buttonText('Next'));
+ *
+ *      it('should show match message if both agree', function() {
+ *        expect(matchMessage.isDisplayed()).toBeFalsy();
+ *        agrees.click();
+ *        expect(matchMessage.isDisplayed()).toBeTruthy();
+ *      });
+ *
+ *      it('should hide match message after next is clicked', function() {
+ *        agrees.click();
+ *        next.click();
+ *        expect(matchMessage.isDisplayed()).toBeFalsy();
+ *      });
+ *   </file>
+ * </example>
+ */
+var ngAsDirective = ['$parse', function($parse) {
+  return {
+    priority: -1,
+    restrict: 'A',
+    compile: function(tElement, tAttrs) {
+      // gets the expected controller name, converts <x-some-thing> into "someThing")
+      var controllerName = tAttrs.$normalize(tElement[0].tagName.toLowerCase());
+
+      // get the setter for the as attribute
+      var getter = $parse(tAttrs.ngAs);
+      var setter = getter.assign;
+
+      return function(scope, iElement, iAttrs) {
+        // gets the controller of the current element (see jqLiteController for details)
+        var controller = iElement.data('$' + controllerName + 'Controller');
+        setter(scope, controller);
+
+        // when the element is removed, remove it from the scope assignment (nullify it)
+        scope.$on('$destroy', function() {
```

Comment:
What if I do it "correctly" but with an isolated scope?

That aside, even if it does mean the developer is doing something wrong... is there an issue with using the element event instead of scope event?